### PR TITLE
Validations for entries

### LINKED ISSUE #13: VALIDATION working in app-spec

### PR DESCRIPTION
**Pull request by lucksus**
closes #13 
This is now about validation working only for entries. Links will be added in a future PR.
Also, I've just defined it here the way it is working right now, which is an intermediary step:
```rust

struct Post {
    content: String,
    date_created: String,
}

pub fn definition() -> ValidatingEntryType {
    entry!(
        name: "post",
        description: "",
        sharing: Sharing::Public,

        validation_package: || {
            hdk::ValidationPackageDefinition::ChainFull
        },

        validation_function: |_entry: Post, _ctx: hdk::ValidationData| {
            Err(String::from("Not in use yet. Will replace validations! macro below soon."))
        }
    )
}

validations! {
    [ENTRY] validate_post {
        |post: Post, _ctx: hdk::ValidationData| {
            (post.content.len() < 280)
                .ok_or_else(|| String::from("Content too long"))
        }
    }
}
```

@sphinxc0re will add the discussed `define_zome!` macro soon and we will replace the `validations!` macro also with dispatch functions as we did for validation_package.

The reason I'm splitting that work into a future PR is to get this closed and only do the change to dispatch functions if there is time left before the devcamp.

### CODE REVIEW

**Comment by lucksus** [comment]
@ddd-mtl can you provide an alternative suggestion then? I would like to implement this tomorrow morning and try having that before the sprint ends..

**Comment by lucksus** [comment]
I had single macros first and worked on coalescing them for better readability. I think custom grammar is way better than putting meta-info in a separate JSON file...

**Comment by ddd-mtl** [comment]
How about this:

File: zomes/blog/code/src/post_validations.rs [hunk 1]
```diff
@@ -1,40 +1,61 @@
...
+        /// like this:
+        ///
+        /// struct ValidationData {
+        ///    sources : Vec<HashString>
+        ///    sourceChainEntries : Option<Vec<serde_json::Value>>
+        ///    sourceChainHeaders : Option<Vec<Header>>
+        ///    custom : Option<serde_json::Value>
+        /// }
+        |post: Post, ctx: hdk::ValidationData| {
+            post.content.len() < 280
```
Comment (lucksus) [inline zomes/blog/code/src/post_validations.rs]:
```rust
if post.content.len() < 280 {
  Ok(())
} else {
  Err("Content too long")
}
```

File: zomes/blog/code/src/post_validations.rs [hunk 2]
```diff
@@ -1,40 +1,89 @@
...
+        /// Should be something like:
+        ///
+        /// enum ValidationPackage {
+        ///    Entry,           //sending only the entry
+        ///    ChainEntries,    //sending all (public?) source chain entries
+        ///    ChainHeaders,    //sending all source chain headers
+        ///    ChainFull,       //sending the whole chain, entries and headers
+        ///    Custom(String),  //sending something custom
+        /// }
+        [ValidationPackage::Entry]
```
Comment (lucksus) [inline zomes/blog/code/src/post_validations.rs]: @zippy: We need to be able to specify certain entry types that should only be included in the package (= not the whole chain).

File: zomes/blog/code/src/post_validations.rs [hunk 3]
```diff
@@ -1,40 +1,89 @@
...
+validations! {
+    /// We also need to distinguish between the validation of an entry
+    /// (here) and validation of links (below)
+    [ENTRY] validate_post {
+        /// This is an enum that needs to be added to the HDK and that tells
+        /// Holochain what validation package the source needs to send
+        /// to validating DHT nodes.
+        /// Should be something like:
+        ///
+        /// enum ValidationPackage {
```
Comment (lucksus) [inline zomes/blog/code/src/post_validations.rs]: -> `enum ValidationPackageRequestType` ?

File: zomes/blog/code/src/lib.rs [hunk 4]
```diff
@@ -1,12 +1,25 @@
...
+use hdk::meta::ZomeDefinition;
+
+/// This function needs to be implemented by every zome.
+/// Every entry type needs to be registered by passing its
+/// definition object (constructed by entry!) to the define
+/// function.
+///
+/// Will soon be replaced by a define_zome! macro
+#[no_mangle]
+pub extern fn zome_setup(zd: &mut ZomeDefinition) {
```
Comment (Connoropolous) [inline zomes/blog/code/src/lib.rs]: I think zd could be expanded, so that it's clearer. Like `zome_defn`

Comment (lucksus) [reply zomes/blog/code/src/lib.rs]: Yes

**Comment by lucksus** [comment]
@Connoropolous, no, we can't get rid of that yet.. We will be able to do so very soon.. but I would (will) create another PR for that which focuses on "no JSON". This one is about validation..

**Comment by Connoropolous** [comment]
ah, right. I did actually discover my assumption about this was wrong late last night. I will switch to approve

File: zomes/blog/code/src/main_capability.rs [hunk 5]
```diff
@@ -13,15 +17,15 @@ zome_functions! {
...
+                let _ = hdk::link_entries(
+                    HashString::from(AGENT_KEY_HASH.to_string()),
                     post_hash.clone(),
                     "authored_posts"
                 );
 
                 let in_reply_to = in_reply_to;
                 if !in_reply_to.to_string().is_empty() {
-                    if let Ok(_) = hdk::get_entry(in_reply_to.clone()) {
+                    if let Ok(_) = hdk::get_entry(in_reply_to.clone(), GetEntryOptions{}) {
```
Comment (sphinxc0re) [inline zomes/blog/code/src/main_capability.rs]: pls use `.is_ok()`

**Pull request merged**